### DISCUSSION TITLE
Confusing error when creating cyclical relationship

### DISCUSSION THREAD

**Discussion by quinnjr** [prisma/prisma#11140] [Q&A] [answered]
Hello,

I'm getting the following error when trying to create a cyclical relationship as the documentation describes:

```sh
Error: Schema parsing
error: Error validating: A self-relation must have `onDelete` and `onUpdate` referential actions set to `NoAction` in one of the @relation attributes. (Implicit default `onDelete`: `SetNull`, and `onUpdate`: `Cascade`) Read more at https://pris.ly/d/cyclic-referential-actions
  -->  schema.prisma:111
   | 
110 |   content   String
111 |   children  Comment? @relation(name: "subcomments", fields: [commentId], references: [id])
112 |   plugin    Plugin   @relation(fields: [pluginId], references: [id])
   | 
```
My prisma versions are:
prima: 3.8.0
@prisma/client: 3.8.0

My model is defined as:
```js
model Comment {
  id             String   @id @default(dbgenerated()) @map("_id") @db.ObjectId
  author         User     @relation(fields: [authorId], references: [id])
  authorId       String   @db.ObjectId
  content        String
  children       Comment[] @relation(name: "subcomments")
  plugin         Plugin   @relation(fields: [pluginId], references: [id])
  pluginId       String   @db.ObjectId
  createdAt      DateTime @default(now()) @db.Timestamp
  updatedAt      DateTime @default(now()) @updatedAt @db.Timestamp

  @@map("comments")
}
```

My configuration for datasource and client configuation is:
```js
datasource db {
  provider = "mongodb"
  url      = env("DATABASE_URL")
  referentialIntegrity = "prisma"
}

generator client {
  provider        = "prisma-client-js"
  previewFeatures = ["mongoDb", "referentialIntegrity"]
  binaryTargets   = ["native"]
}
```

I'm confused on where I need to change something to get the relation to work.

**Comment by janpio** [comment, marked as answer]
The error message shows a different model than the one you posted. Is that indeed correct?

**Reply by quinnjr** [reply]
> The error message shows a different model than the one you posted. Is that indeed correct?

Yes, the model is different in the error message as compared to what is defined in the schema.

**Reply by janpio** [reply]
Ok, I think I get it now:

You write `children       Comment[] @relation(name: "subcomments")` in your schema, but you get an error message when running `npx prisma format` to fix the error that shows, and that error message then says `children  Comment? @relation(name: "subcomments", fields: [commentId], references: [id])` in the output with the "A self-relation must have..." error message. Correct?

**Reply by janpio** [reply]
> I'm getting the following error when trying to create a cyclical relationship as the documentation describes:

Can you share the link of the documentation you are trying to follow? Thanks.

**Reply by quinnjr** [reply]
> > I'm getting the following error when trying to create a cyclical relationship as the documentation describes:
> 
> Can you share the link of the documentation you are trying to follow? Thanks.

I'm following the documentation at https://www.prisma.io/docs/concepts/components/prisma-schema/relations/referential-actions/special-rules-for-referential-actions

**Reply by quinnjr** [reply]
> Ok, I think I get it now:
> 
> You write `children Comment[] @relation(name: "subcomments")` in your schema, but you get an error message when running `npx prisma format` to fix the error that shows, and that error message then says `children Comment? @relation(name: "subcomments", fields: [commentId], references: [id])` in the output with the "A self-relation must have..." error message. Correct?

Correct

**Reply by janpio** [reply]
> I'm following the documentation at [prisma.io/docs/concepts/components/prisma-schema/relations/referential-actions/special-rules-for-referential-actions](https://www.prisma.io/docs/concepts/components/prisma-schema/relations/referential-actions/special-rules-for-referential-actions)

I think you are missing the other side of the relation then. It could look like this:

```
  parentId String
  parent   Comment?  @relation(name: "children", fields: [parentId], references: [id])
  children Comment[] @relation(name: "children")
```

You will still get the error message, but then be able to fix it to this:
```
  parentId String
  parent   Comment?  @relation(name: "children", fields: [parentId], references: [id], onUpdate: NoAction)
  children Comment[] @relation(name: "children")
```

Does that work for you?

**Reply by janpio** [reply]
Written up as bug report now in https://github.com/prisma/prisma/issues/11176

**Reply by quinnjr** [reply]
> > I'm following the documentation at [prisma.io/docs/concepts/components/prisma-schema/relations/referential-actions/special-rules-for-referential-actions](https://www.prisma.io/docs/concepts/components/prisma-schema/relations/referential-actions/special-rules-for-referential-actions)
> 
> I think you are missing the other side of the relation then. It could look like this:
> 
> ```
>   parentId String
>   parent   Comment?  @relation(name: "children", fields: [parentId], references: [id])
>   children Comment[] @relation(name: "children")
> ```
> 
> You will still get the error message, but then be able to fix it to this:
> 
> ```
>   parentId String
>   parent   Comment?  @relation(name: "children", fields: [parentId], references: [id], onUpdate: NoAction)
>   children Comment[] @relation(name: "children")
> ```
> 
> Does that work for you?

Yes, this fixes the issue with `pnpx prisma format`

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
